src/main/java/io/divolte/server/filesinks/gcs/GoogleCloudStorageFileManager.java

```diff
@@ -0,0 +1,406 @@
+    }
+
+    public class GoogleCloudStorageDivolteFile implements DivolteFile {
+        private final AvroRecordBuffer[] buffer;
+        private final DataFileWriter<GenericRecord> writer;
+        private final DynamicDelegatingOutputStream avroTargetStream;
+
+        private final String inflightNameEncoded;
+        private final String inflightPartialNameEncoded;
+        private final String publishNameEncoded;
+
+        private final String inflightName;
+        private final String inflightPartialName;
+
+        private boolean partWritten;
+        private int position;
+
+        private GoogleCloudStorageDivolteFile(final String fileName) throws IOException {
+            /*
+             * Consider pooling these or assume only one file to be active at any point in
+             * time and use a single buffer per outer instance. While the latter is
+             * currently valid, new file syncing and rolling strategies might change this.
+             */
+            this.buffer = new AvroRecordBuffer[recordBufferSize];
+
+            this.inflightName = inflightDir + GCS_PATH_SEPARATOR_CHAR + fileName;
+            this.inflightPartialName = inflightName + PART_CLASSIFIER;
+
+            this.inflightNameEncoded =  URLEncoder.encode(inflightName, URL_ENCODING);
+            this.inflightPartialNameEncoded = inflightNameEncoded + PART_CLASSIFIER;
+            this.publishNameEncoded =  URLEncoder.encode(publishDir + GCS_PATH_SEPARATOR_CHAR + fileName, URL_ENCODING);
+
+            final URL remoteFileUrl = uploadUrlFor(bucketEncoded, inflightNameEncoded);
+
+            final HttpURLConnection connection = setupUrlConnection(POST, remoteFileUrl, true, AVRO_CONTENT_TYPE_HEADER);
+            final OutputStream os = connection.getOutputStream();
+
+            /*
+             * We create a single Avro writer, but write parts of the Avro stream to
+             * multiple files, which are composed into a single file after flushing files.
+             * We use a DynamicDelegatingOutputStream for this, which is a output stream
+             * wrapper that supports changing the wrapped stream on the fly.
+             *
+             * When creating a Avro writer, it immediately writes the Avro header to the
+             * underlying stream. As such, we need to open the HTTP connection before
+             * creating the writer. As the writer is final, we cannot use the
+             * googlePost(...) helper in the constructor, because we cannot set a final from
+             * a lambda.
+             */
+            avroTargetStream = new DynamicDelegatingOutputStream();
+            avroTargetStream.attachDelegate(os);
+            writer = new DataFileWriter<GenericRecord>(new GenericDatumWriter<>(schema)).create(schema, avroTargetStream);
+            writer.flush();
+            avroTargetStream.detachDelegate();
+
+            os.flush();
+            os.close();
```

Comment:
Removed the flush as detaching already flushes indeed. I don't know about the flush on close behaviour, but don't want to make any assumptions.